### PR TITLE
Laplace approximation-based uncertainty estimation

### PR DESCRIPTION
**Pull request by joshuaeyu**
Additions: 
- `NSM/uncertainty.py`: Helper functions and classes for uncertainty estimation. The latent posterior is approximated using Laplace approximation, and SDF uncertainty is estimated using either local linearization analysis or Monte Carlo sampling to propagate the latent covariance through the model.
- `uncertainty_demo.py`: A notebook demonstrating shape completion with uncertainty estimation (and optional uncertainty hyperparameter tuning).

Changes:
- `NSM/optimization.py`: Add support for L2 data loss and use PyTorch's builtin `F.mse_loss` for it and regularization loss.

### CODE REVIEW

**Review by aubricot — APPROVED** [review]
accept for now, but refactor shape_completion.py and shape_completion_slicer.py using updates in uncertainty.py and push funcs to ext module
